### PR TITLE
Check if a double is NaN before converting to it int

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

`RenderObject.localToGlobal` may return `(NaN, NaN)` when `(x, y, 0, 1)` is transformed to a point at infinity. This is generally not a problem, as painting is usually skipped when the transform is set to zero. But trying to convert the point's coordinates to int or json will throw.

## Related Issues

https://github.com/flutter/flutter/issues/9344

## Tests

I added the following tests:

- getLocalRectForCaret does not throw when it sees an infinite point

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/painting/matrix_utils.dart [hunk 1]
```diff
@@ -127,6 +127,13 @@ class MatrixUtils {
   ///
   /// This function assumes the given point has a z-coordinate of 0.0. The
   /// z-coordinate of the result is ignored.
+  ///
+  /// While not common, this method may return (NaN, NaN), iff the given `point`
```
Comment (Piinks) [inline packages/flutter/lib/src/painting/matrix_utils.dart:131]: Just checking - 'iff' as in 'if and only if', or typo?

**Review by Piinks — APPROVED** [review]
LGTM!

File: packages/flutter/lib/src/painting/matrix_utils.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/matrix_utils.dart:131]: Yeah I meant "if and only if".

**Review by justinmc — APPROVED** [review]
LGTM

I ran into something vaguely similar but not quite the same while working with matrix math in InteractiveViewer.  Some Matrix4 methods would fail in strange ways when given infinite numbers.  I'm glad this will handle things in a more sane way now at least 👍

**Pull request merged**